### PR TITLE
Json generic feed

### PR DESCRIPTION
**Pull request by teizenman**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/21755#event-3096699755

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [X] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    params['feed_name_to_config'] = {
+        'JSON': {
+            'url': params.get('url'),
+            'extractor': params.get('extractor'),
+            'indicator': params.get('indicator'),
+            'indicator_type': params.get('indicator_type'),
+        }
+    }
+
+    feed_main(params, 'JSONFeed', 'json')
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.py]:
```suggestion
    feed_main(params, 'JSON Feed', 'json')
```

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 2]
```diff
@@ -0,0 +1,125 @@
+category: Data Enrichment & Threat Intelligence
+commonfields:
+  id: JSONFeed
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]:
```suggestion
  id: JSON Feed
```

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  - C - Fairly reliable
+  - D - Not usually reliable
+  - E - Unreliable
+  - F - Reliability cannot be judged
+  required: true
+  type: 15
+  additionalinfo: Reliability of the source providing the intelligence data
+- display: ""
+  name: feedExpirationPolicy
+  defaultvalue: suddenDeath
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]: Why is this the default?

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+- additionalinfo: Type of the indicator in the feed.
+  display: Indicator Type
+  name: indicator_type
+  required: false
+  type: 0
+- display: Username
+  name: credentials
+  required: false
+  type: 9
+- additionalinfo: JMESPath expression for extracting the indicators from.
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]:
```suggestion
- additionalinfo: JMESPath expression for extracting the indicators.
```

Comment (avidan-H) [reply Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]: Please include additional information here in the tooltip, e.g. a link to JMESPath expression syntax perhaps and/or an example.

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  display: JMESPath Extractor
+  name: extractor
+  required: false
+  type: 0
+- additionalinfo: JSON attribute to use as indicator. Default is 'indicator'.
+  display: JSON Attribute Indicator
+  name: indicator
+  required: false
+  type: 0
+- additionalinfo: The names to apply the fields in the JSON feed.
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]:
```suggestion
- additionalinfo: The names to apply to the fields in the JSON feed.
```

Comment (avidan-H) [reply Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]: This tooltip here doesn't sufficiently explain what this parameter is. Please make it more detailed/informative.

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  required: false
+  type: 0
+- display: Username
+  name: credentials
+  required: false
+  type: 9
+- additionalinfo: JMESPath expression for extracting the indicators from.
+  display: JMESPath Extractor
+  name: extractor
+  required: false
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]: For this parameter, and the following parameters, `required` is `false`. What happens if I don't include any of these fields?

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  type: 0
+- additionalinfo: The names to apply the fields in the JSON feed.
+  display: Field Names
+  name: fields
+  required: false
+  type: 0
+description: Fetch indicators from a JSON feed.
+display: JSON Feed
+name: JSONFeed
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]:
```suggestion
name: JSON Feed
```

**Comment by teizenman** [comment]
@avidan-H Fixed, had to push it like that with the old sdk version

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+Fetch indicators from a JSON feed. The integration allows a great amount of user configuration to support different types of JSON feeds.
+
+* **URL** - URL of the feed.
+* **Indicator Type** - The type of indicators in the feed.
+* **Username + Password** - Credentials to access feeds that require basic authentication. 
+These fields also support the use of API key headers. To use API key headers, specify the header name and value in the following format:
+`_header:<header_name>` in the **Username** field and the header value in the **Password** field.
+* **JMESPath Extractor** - JMESPath expression for extracting the indicators from. You can check the expression in 
+the [JMESPath site](http://jmespath.org/) to verify this expression will return the following array of objects.
+* **JSON Attribute Indicator** - JSON attribute to use as indicator. Default is 'indicator'.
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md]:
```suggestion
* **JSON Indicator Attribute** - JSON attribute whose value is the indicator. Default is 'indicator'.
```

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  name: credentials
+  required: false
+  type: 9
+- additionalinfo: JMESPath expression for extracting the indicators from.
+  display: JMESPath Extractor
+  name: extractor
+  required: false
+  type: 0
+- additionalinfo: JSON attribute to use as indicator. Default is 'indicator'.
+  display: JSON Attribute Indicator
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]:
```suggestion
  display: JSON Indicator Attribute
```

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+**Indicator Type** - CIDR.
+
+**Credentials** - This feed does not require authentication.
+
+From a quick look at the feed in the web browser, we are going to configure the rest of the parameters:
+
+**JMESPath Extractor** - prefixes[?service=='AMAZON'].
+
+**JSON attribute indicator** - ip_prefix.
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md]:
```suggestion
**JSON Indicator Attribute** - ip_prefix.
```

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+`_header:<header_name>` in the **Username** field and the header value in the **Password** field.
+* **JMESPath Extractor** - JMESPath expression for extracting the indicators from. You can check the expression in 
+the [JMESPath site](http://jmespath.org/) to verify this expression will return the following array of objects.
+* **JSON Attribute Indicator** - JSON attribute to use as indicator. Default is 'indicator'.
+* **Field Names** - The names to apply the fields in the JSON feed.
+
+## Step by step configuration
+As an example, we'll be looking at the IP ranges from Amazon AWS. This feed will ingest indicators of type CIDR. These are the feed instance configuration parameters for our example.
+
+**URL**: https://ip-ranges.amazonaws.com/ip-ranges.json.
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md]: If you are using plaintext as the example values for all of these configuration step parameters - then please remove the closing periods as someone could get confused and accidentally include one of them.

File: Tests/conf.json [hunk 12]
```diff
@@ -2,6 +2,11 @@
     "testTimeout": 160,
     "testInterval": 20,
     "tests": [
+        {
+            "integrations": "JSONFeed",
```
Comment (avidan-H) [inline Tests/conf.json]:
```suggestion
            "integrations": "JSON Feed",
```

File: Packs/FeedJSON/TestPlaybooks/JSON_Feed_Test.yml [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+      "dimensions": {
+        "height": 735,
+        "width": 380,
+        "x": 50,
+        "y": 50
+      }
+    }
+  }
+inputs: []
+outputs: []
```
Comment (avidan-H) [inline Packs/FeedJSON/TestPlaybooks/JSON_Feed_Test.yml]: Please add `fromversion`.

**Comment by avidan-H** [comment]
After you take care of these fixes, let's sit down for a demo session.

**Comment by teizenman** [comment]
@kirbles19 Can you please go over the following files:

- https://github.com/demisto/content/blob/json-generic-feed/Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml

- https://github.com/demisto/content/blob/json-generic-feed/Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md

File: Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py [hunk 14]
```diff
@@ -155,7 +155,7 @@ def get_feed_config(self, fields_json: str = '', indicator_json: str = ''):
         :return: The feed configuration.
         """
         config = {}
-        if indicator_json:
+        if indicator_json is not None:
```
Comment (avidan-H) [inline Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py]: Why did you revert the change from master? This is no bueno.

File: Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule_test.py [hunk 15]
```diff
@@ -137,17 +137,3 @@ def test_datestring_to_millisecond_timestamp():
     assert 1581341954000 == datestring_to_millisecond_timestamp(datesting5)
     assert 1581341954123 == datestring_to_millisecond_timestamp(datesting3)
     assert 1581341954123 == datestring_to_millisecond_timestamp(datesting4)
-
```
Comment (avidan-H) [inline Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule_test.py]: Same here...

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 16]
```diff
@@ -154,6 +139,75 @@ def fetch_indicators_command(client: Client, indicator_type: str, **kwargs) -> U
...
+
+    if re.match(sha1Regex, indicator_value):
+        return FeedIndicatorType.File
+
+    if re.match(emailRegex, indicator_value):
+        return FeedIndicatorType.Email
+
+    try:
+        if tldextract.extract(indicator_value).suffix:
+            return FeedIndicatorType.Domain
```
Comment (avidan-H) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: You didn't include the `DomainGlob`.

File: Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py [hunk 17]
```diff
@@ -154,6 +139,75 @@ def fetch_indicators_command(client: Client, indicator_type: str, **kwargs) -> U
...
+                elif key != indicator_key:
+                    fields[key] = value
+        elif json_element and indicator_key not in json_element:
+            fields.update(json_element)
+
+    extract(indicator)
+    return fields
+
+
+def auto_detect_indicator_type(indicator_value):
```
Comment (avidan-H) [inline Packs/ApiModules/Scripts/JSONFeedApiModule/JSONFeedApiModule.py]: Please move these helper functions to the top.

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+        params.get('url'): {
+            'url': params.get('url'),
+            'extractor': params.get('extractor'),
+            'indicator': params.get('indicator', 'indicator'),
+        }
+    }
+
+    if not params.get('auto_detect_type'):
+        if not params.get('indicator_type'):
+            return_error('Indicator Type can not be emoty when Auto Detect Indicator Type is unchecked')
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.py]:
```suggestion
            return_error('Indicator Type cannot be empty when Auto Detect Indicator Type is unchecked')
```

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+- additionalinfo: The JSON attribute that holds the indicator value. Default value is 'indicator'.
+  display: JSON Indicator Attribute
+  name: indicator
+  required: false
+  type: 0
+- additionalinfo: A comma seperated list of fields to to extract from a JSON object.
+  display: Field Names
+  name: fields
+  required: false
+  type: 0
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]: Didn't you remove `fields` from the code?

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+  defaultvalue: "20160"
+  required: false
+  type: 1
+  display: ""
+- defaultvalue: '240'
+  display: Feed Fetch Interval
+  name: feedFetchInterval
+  required: false
+  type: 19
+- display: Bypass exclusion list
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]: I don't like all these checkboxes interspersed around the configuration panel parameters. Put them all together at the bottom please.

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+  required: false
+  additionalinfo:
+    When selected, the exclusion list is ignored for indicators from this feed.
+    This means that if an indicator from this feed is on the exclusion list,
+    the indicator might still be added to the system.
+- display: URL
+  name: url
+  required: true
+  type: 0
+- display: Auto detect indicator type
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON.yml]: The same as above.

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+* **URL** - URL of the feed.
+* **Auto detect indicator type** - If checked, a type auto detection mechanism will take place for each indicator.
+* **Indicator Type** - Type of the indicator in the feed. Relevant only if Auto detect is not checked.
+* **Username + Password** - Credentials to access feeds that require basic authentication. 
+These fields also support the use of API key headers. To use API key headers, specify the header name and value in the following format:
+`_header:<header_name>` in the **Username** field and the header value in the **Password** field.
+* **JMESPath Extractor** - JMESPath expression for extracting the indicators from. You can check the expression in 
+the [JMESPath site](http://jmespath.org/) to verify this expression will return the following array of objects.
+* **JSON Indicator Attribute** - JSON attribute whose value is the indicator. Default is 'indicator'.
+* **Field Names** - The names to apply the fields in the JSON feed.
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md]: Same question - didn't you remove `fields` from the parameters?

File: Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+
+From a quick look at the feed in the web browser, we are going to configure the rest of the parameters:
+
+**JMESPath Extractor** - prefixes[?service=='AMAZON'] This means that the desired objects to extract the indicators from is
+`prefixes`, and the objects will be filtered by where the field `service` is equal to `AMAZON`.
+
+**JSON Indicator Attribute** - ip_prefix
+
+**Field Names** - We have 2 fields in this feed - `region,service`. The integration ignores these headers and we have to configure the field names for each indicator.
+So we will configure these field names: `region,service`, so that the indicator will be created with these fields.
```
Comment (avidan-H) [inline Packs/FeedJSON/Integrations/FeedJSON/FeedJSON_description.md]: Same question as above.

File: Packs/FeedJSON/TestPlaybooks/JSON_Feed_Test.yml (hunk 13 above)
Comment (avidan-H) [reply Packs/FeedJSON/TestPlaybooks/JSON_Feed_Test.yml]: You still haven't added a `fromversion` field unless I missed it.

File: Tests/conf.json [hunk 24]
```diff
@@ -2,6 +2,17 @@
     "testTimeout": 160,
     "testInterval": 20,
     "tests": [
+        {
+            "integrations": ["Syslog Sender", "syslog"],
+            "playbookID": "Test Syslog",
+            "fromversion": "5.5.0",
+            "timeout": 1000
+        },
```
Comment (avidan-H) [inline Tests/conf.json]: I don't understand why you have `Syslog` here.

**Comment by kirbles19** [comment]
@teizenman Done.
